### PR TITLE
Connection events from conntrack

### PR DESCRIPTION
**Pull request by iBigQ**
This adds a new event publisher named `conntrack` that uses the netfilter subsystem of netlink to retrieve status changes of network connections. 
The table `connection_events` tags connections (IP and port combinations of source and destination) with labels NEW, UPDATE or DESTROYED. Connections are matched with /proc/net/<protocol> to find the corresponding inode and /proc/<pid>/fd is browsed to find the respective PID.

### CODE REVIEW

**Comment by iBigQ** [comment]
The `conntrack` event contains the network information only, and the PID is detected **after** the event happened. Thus there is a possibility that either the socket is already closed (no inode from /proc/net) or the process already died (no pid from /proc/<pid>/fd).
This e.g. can happen for short-lived `curl` processes. The `connection_events` table then reports the network connection information only.

**Comment by iBigQ** [comment]
TODO: The formulas create both static and dynamic libraries (we need static only)

File: osquery/events/linux/conntrack.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#include <osquery/logger.h>
+#include <osquery/system.h>
+
+#include "osquery/events/linux/conntrack.h"
+
+namespace osquery {
+
+/// The conntrack subsystem may have a performance impact on the system.
+FLAG(bool,
+     disable_conntrack,
```
Comment (theopolis) [inline osquery/events/linux/conntrack.cpp]: If this is default-disabled, let's rename to `enable_conntrack` and set the default to `false`. Then if folks want the capability they'll use: `--enable_conntrack`.

File: osquery/events/linux/conntrack.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    return Status(1, "Publisher disabled via configuration");
+  }
+
+  nl_ = std::shared_ptr<struct mnl_socket>(mnl_socket_open(NETLINK_NETFILTER),
+                                           mnl_socket_close);
+  if (nl_ == nullptr) {
+    return Status(1, "Could not open conntrack subsystem");
+  }
+
+  // TODO: How to test if kernel modules have been loaded?
```
Comment (theopolis) [inline osquery/events/linux/conntrack.cpp]: Still some TODOs, let's create issues for them, implement them, or remove the line.

File: osquery/events/linux/conntrack.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  if (mnl_socket_bind(nl_.get(),
+                      NF_NETLINK_CONNTRACK_NEW | NF_NETLINK_CONNTRACK_UPDATE |
+                          NF_NETLINK_CONNTRACK_DESTROY,
+                      MNL_SOCKET_AUTOPID) < 0) {
+    return Status(1, "Could not subscribe to updates from conntrack subsystem");
+  }
+
+  return Status(0, "OK");
+}
+/**
```
Comment (theopolis) [inline osquery/events/linux/conntrack.cpp]: nit, space between symbol and docblock.

File: osquery/events/linux/conntrack.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ * @return netlink success code
+ */
+static int data_cb(const struct nlmsghdr* nlh, void* data) {
+  ConntrackEventContext* ec = reinterpret_cast<ConntrackEventContext*>(data);
+  struct nf_conntrack* ct = ec->event.get();
+  enum nf_conntrack_msg_type type = NFCT_T_UNKNOWN;
+
+  switch (nlh->nlmsg_type & 0xFF) {
+  case IPCTNL_MSG_CT_NEW:
+    if (nlh->nlmsg_flags & (NLM_F_CREATE | NLM_F_EXCL))
```
Comment (theopolis) [inline osquery/events/linux/conntrack.cpp]: nit, need `{` around conditional blocks.

File: osquery/events/linux/conntrack.cpp [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    else
+      type = NFCT_T_UPDATE;
+    break;
+  case IPCTNL_MSG_CT_DELETE:
+    type = NFCT_T_DESTROY;
+    break;
+  }
+  ec->type = type;
+
+  if (ct == NULL)
```
Comment (theopolis) [inline osquery/events/linux/conntrack.cpp]: Prefer `nullptr` and use `{` around all blocks.

File: osquery/events/linux/conntrack.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    return MNL_CB_OK;
+
+  nfct_nlmsg_parse(nlh, ct);
+  /**
+  char buf[4096];
```
Comment (theopolis) [inline osquery/events/linux/conntrack.cpp]: nit, please remove dead code.

File: osquery/events/linux/conntrack.h [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  // TODO: Any subscription specific parameters? E.g. event types?
+};
+
+/**
+ * @brief Event details for ConntrackEventPublisher events.
+ */
+struct ConntrackEventContext : public EventContext {
+  /// The nf_conntrack structure if the EventSubscriber want to interact.
+  std::shared_ptr<struct nf_conntrack> event{nullptr};
+  /// A nf_conntrack_msg_type action representing the event type (NEW, UPDATE,
```
Comment (theopolis) [inline osquery/events/linux/conntrack.h]: If the short-hand docblock needs more than 1-line use the long-doc version.

File: osquery/events/linux/conntrack.h [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+#include <libmnl/libmnl.h>
+#include <libnetfilter_conntrack/libnetfilter_conntrack.h>
+
+#include <osquery/events.h>
+
+namespace osquery {
+
+/**
+ * @brief Subscription details for ConntrackEventPublisher events.
```
Comment (theopolis) [inline osquery/events/linux/conntrack.h]: If you're docblock only needs 1-line, use the short-doc version `///`

File: osquery/events/linux/conntrack.h [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+ * netlink.
+ * Conntrack tracks the connection status of network flows.
+ */
+class ConntrackEventPublisher
+    : public EventPublisher<ConntrackSubscriptionContext,
+                            ConntrackEventContext> {
+  DECLARE_PUBLISHER("conntrack");
+
+ public:
+  virtual ~ConntrackEventPublisher(){};
```
Comment (theopolis) [inline osquery/events/linux/conntrack.h]: nit, space between () and {}.

File: osquery/tables/networking/utils.h [hunk 10]
```diff
@@ -31,5 +34,34 @@ std::string ipAsString(const struct in_addr *in);
 std::string macAsString(const struct ifaddrs *addr);
 std::string macAsString(const char *addr);
 int netmaskFromIP(const struct sockaddr *in);
+
+// Linux proc protocol define to net stats file name.
+extern const std::map<int, std::string> kLinuxProtocolNames;
+// A map of socket handles (inodes) to their pid and file descriptor.
+typedef std::map<std::string, std::pair<std::string, std::string>> InodeMap;
+std::string addressFromHex(const std::string& encoded_address, int family);
+unsigned short portFromHex(const std::string& encoded_port);
```
Comment (theopolis) [inline osquery/tables/networking/utils.h]: I know we haven't set a good example in this header but do you mind include spaces between the symbol declarations. And if possible using the `///` short-doc docblock if it doesn't need param and return value documentation?

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 9449c64a5e7ae7269a091fa5e2a267e44868c365 (Job results: [1878](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1878/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 9449c64a5e7ae7269a091fa5e2a267e44868c365 (Job results: [1879](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1879/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 9449c64a5e7ae7269a091fa5e2a267e44868c365 (Job results: [5769](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5769/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 9449c64a5e7ae7269a091fa5e2a267e44868c365 (Job results: [698](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/698/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 9449c64a5e7ae7269a091fa5e2a267e44868c365 (Job results: [5770](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5770/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 9449c64a5e7ae7269a091fa5e2a267e44868c365 (Job results: [699](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/699/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
Build finished.

**Comment by osqueryer** [comment]
Build finished. 
:no_good: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [2777](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2777/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [2777](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2777/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [2250](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2250/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [2250](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2250/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [6148](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6148/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [6148](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6148/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [3026](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3026/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [3026](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3026/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [3027](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3027/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [3027](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3027/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [2251](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2251/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [2251](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2251/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
Build finished. 
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [6149](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6149/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit a1178c7b3d857e1680ff1c6f36ed794b97be2606 (Job results: [6149](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6149/)) failed one or more tests (**Linux**).

**Pull request closed without merging**